### PR TITLE
TRUNK-5208:ModuleUtil compareVersions should sort a SNAPSHOT version as "less than"

### PR DESCRIPTION
**Pull request by jwnasambu**
## Description of what I changed
I replaced the custom version ordering code with [SemVer ](https://github.com/vdurmont/semver4j)

## Issue I worked on
https://issues.openmrs.org/browse/TRUNK-5208

### CODE REVIEW

**Comment by HerbertYiga** [comment]
@jwamalwa Thanks for working on this, i see some breaking point ,i picked this line from your travis logs ([ERROR] /home/travis/build/openmrs/openmrs-core/api/src/main/java/org/openmrs/module/ModuleUtil.java:[72,85] illegal character: '\')

File: api/src/main/java/org/openmrs/module/ModuleUtil.java [hunk 1]
```diff
@@ -67,6 +67,11 @@ private ModuleUtil() {
 	 *
 	 * @param props Properties (OpenMRS runtime properties)
 	 */
+	// adapted from semver.org
+	private static final Pattern VERSION_PATTERN = Pattern.compile(
+		""^(?<major>0|[1-9]\\d*)\\.(?<minor>0|[1-9]\\d*)\\.(?<patch>0|[1-9]\\d*)(?:-(?<snapshot>SNAPSHOT))?$""
+	);
+
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/ModuleUtil.java:74]: This block should come before the Javadoc string.

File: api/src/main/java/org/openmrs/module/ModuleUtil.java [hunk 2]
```diff
@@ -471,8 +488,26 @@ public static int compareVersion(String version, String value) {
...
+		final String[] versions = {
+			versionMatcher.group("major"),
+			versionMatcher.group("minor"),
+			versionMatcher.group("patch")
+		};
+		final String[] values = {
+			valueMatcher.group("major"),
+			valueMatcher.group("minor"),
+			valueMatcher.group("patch")
+		};
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/ModuleUtil.java:510]: Sorry, this block was intended to replace lines 426-450 above (the part in `compareVersion()` that is breaking the string down into it's component parts). The loop from lines 452-454 will also need to be adjusted.

File: api/src/main/java/org/openmrs/module/ModuleUtil.java [hunk 3]
```diff
@@ -62,11 +62,17 @@ private ModuleUtil() {
 	
 	private static final Logger log = LoggerFactory.getLogger(ModuleUtil.class);
 	
+	// adapted from semver.org
+		private static final Pattern VERSION_PATTERN = Pattern.compile(
+			""^(?<major>0|[1-9]\\d*)\\.(?<minor>0|[1-9]\\d*)\\.(?<patch>0|[1-9]\\d*)(?:-(?<snapshot>SNAPSHOT))?$""
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/ModuleUtil.java:67]:
```suggestion
			"^(?<major>0|[1-9]\\d*)\\.(?<minor>0|[1-9]\\d*)\\.(?<patch>0|[1-9]\\d*)(?:-(?<snapshot>SNAPSHOT))?$"
```

You've got two sets of double quotes there, which doesn't work very well.

File: api/src/main/java/org/openmrs/module/ModuleUtil.java [hunk 4]
```diff
@@ -418,31 +424,24 @@ public static int compareVersion(String version, String value) {
...
+				versionMatcher.group("minor"),
+				versionMatcher.group("patch")
+			};
+			final String[] values = {
+				valueMatcher.group("major"),
+				valueMatcher.group("minor"),
+				valueMatcher.group("patch")
+			};
 			
 			for (int x = 0; x < versions.size(); x++) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/ModuleUtil.java:446]: So now that we've changed `versions` and `values` from lists to arrays, we need adapt this look to use `versions.length` and, e.g., `versions[x]` to get teh value.

**Comment by jwnasambu** [comment]
@ibacher  thanks for the review I have fixed the proposed changes.

**Comment by jwnasambu** [comment]
@ibacher I kindly need your guidance on how to solve this error https://pastebin.com/QAKV87MC I tried fixing it on weekend but all in vain.

**Pull request closed without merging**